pkg/sidecar-controller/framework_test.go

```diff
@@ -0,0 +1,1025 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sidecar_controller
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	crdv1 "github.com/kubernetes-csi/external-snapshotter/pkg/apis/volumesnapshot/v1beta1"
+	clientset "github.com/kubernetes-csi/external-snapshotter/pkg/client/clientset/versioned"
+	"github.com/kubernetes-csi/external-snapshotter/pkg/client/clientset/versioned/fake"
+	snapshotscheme "github.com/kubernetes-csi/external-snapshotter/pkg/client/clientset/versioned/scheme"
+	informers "github.com/kubernetes-csi/external-snapshotter/pkg/client/informers/externalversions"
+	storagelisters "github.com/kubernetes-csi/external-snapshotter/pkg/client/listers/volumesnapshot/v1beta1"
+	"github.com/kubernetes-csi/external-snapshotter/pkg/utils"
+	v1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/diff"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	kubefake "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/kubernetes/scheme"
+	core "k8s.io/client-go/testing"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog"
+)
+
+// This is a unit test framework for snapshot controller.
+// It fills the controller with test snapshots/contents and can simulate these
+// scenarios:
+// 1) Call syncSnapshot/syncContent once.
+// 2) Call syncSnapshot/syncContent several times (both simulating "snapshot/content
+//    modified" events and periodic sync), until the controller settles down and
+//    does not modify anything.
+// 3) Simulate almost real API server/etcd and call add/update/delete
+//    content/snapshot.
+// In all these scenarios, when the test finishes, the framework can compare
+// resulting snapshots/contents with list of expected snapshots/contents and report
+// differences.
+
+// controllerTest contains a single controller test input.
+// Each test has initial set of contents and snapshots that are filled into the
+// controller before the test starts. The test then contains a reference to
+// function to call as the actual test. Available functions are:
+//   - testSyncSnapshot - calls syncSnapshot on the first snapshot in initialSnapshots.
+//   - testSyncSnapshotError - calls syncSnapshot on the first snapshot in initialSnapshots
+//                          and expects an error to be returned.
+//   - testSyncContent - calls syncContent on the first content in initialContents.
+//   - any custom function for specialized tests.
+// The test then contains list of contents/snapshots that are expected at the end
+// of the test and list of generated events.
```

Comment:
These comments need to be cleaned up.  It talks about initialSnapshots and syncSnapshot, etc.